File: lib/presentation/widgets/price_tag.dart
import 'package:flutter/material.dart';

/// A price label with optional struck-through original price and a
/// discount badge.
class PriceTag extends StatelessWidget {
  const PriceTag({
    super.key,
    required this.price,
    this.oldPrice,
    this.currency = '\$',
  });

  final double price;
  final double? oldPrice;
  final String currency;

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final hasDiscount = oldPrice != null && oldPrice! > price;
    final percent =
        hasDiscount ? (((oldPrice! - price) / oldPrice!) * 100).round() : 0;

    return Row(
      mainAxisSize: MainAxisSize.min,
      crossAxisAlignment: CrossAxisAlignment.end,
      children: [
        Text(
          '$currency${price.toStringAsFixed(2)}',
          style: TextStyle(
            fontSize: 24,
            fontWeight: FontWeight.bold,
            color: scheme.primary,
          ),
        ),
        if (hasDiscount) ...[
          const SizedBox(width: 8),
          Padding(
            padding: const EdgeInsets.only(bottom: 3),
            child: Text(
              '$currency${oldPrice!.toStringAsFixed(2)}',
              style: TextStyle(
                fontSize: 15,
                color: scheme.onSurfaceVariant,
                decoration: TextDecoration.lineThrough,
              ),
            ),
          ),
          const SizedBox(width: 8),
          Container(
            padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 3),
            margin: const EdgeInsets.only(bottom: 3),
            decoration: BoxDecoration(
              color: scheme.errorContainer,
              borderRadius: BorderRadius.circular(6),
            ),
            child: Text(
              '-$percent%',
              style: TextStyle(
                color: scheme.onErrorContainer,
                fontWeight: FontWeight.bold,
                fontSize: 12,
              ),
            ),
          ),
        ],
      ],
    );
  }
}

File: android_native/app/src/main/java/com/example/widgetgallery/presentation/components/PriceTag.kt
package com.example.widgetgallery.presentation.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlin.math.roundToInt

/** Price with optional old price and a discount badge. */
@Composable
fun PriceTag(
    price: Double,
    modifier: Modifier = Modifier,
    oldPrice: Double? = null,
    currency: String = "$",
) {
    val scheme = MaterialTheme.colorScheme
    val hasDiscount = oldPrice != null && oldPrice > price
    val percent = if (hasDiscount)
        (((oldPrice!! - price) / oldPrice) * 100).roundToInt() else 0
    Row(modifier, verticalAlignment = Alignment.Bottom) {
        Text("$currency${"%.2f".format(price)}",
            fontSize = 24.sp, fontWeight = FontWeight.Bold,
            color = scheme.primary)
        if (hasDiscount) {
            Spacer(Modifier.width(8.dp))
            Text("$currency${"%.2f".format(oldPrice)}",
                fontSize = 15.sp, color = scheme.onSurfaceVariant,
                textDecoration = TextDecoration.LineThrough,
                modifier = Modifier.padding(bottom = 3.dp))
            Spacer(Modifier.width(8.dp))
            Box(
                Modifier
                    .padding(bottom = 3.dp)
                    .clip(RoundedCornerShape(6.dp))
                    .background(scheme.errorContainer)
                    .padding(horizontal = 8.dp, vertical = 3.dp),
            ) {
                Text("-$percent%", color = scheme.onErrorContainer,
                    fontWeight = FontWeight.Bold, fontSize = 12.sp)
            }
        }
    }
}
